lib/JSON/Validator.pm

```diff
@@ -155,9 +155,10 @@ sub joi {
 
 sub load_and_validate_schema {
   my ($self, $spec, $args) = @_;
-  my $schema = $args->{schema} || SPECIFICATION_URL;
-  $self->version($1) if !$self->{version} and $schema =~ /draft-0+(\w+)/;
+  my $schema = $args->{schema} || (ref $spec eq 'HASH' && $spec->{'$schema'});
+  $self->version($1) if $schema and !$self->{version} and $schema =~ /draft-0+(\w+)/;
   $spec = $self->_resolve($spec);
+  $schema ||= $spec->{'$schema'} || SPECIFICATION_URL;
```

Comment:
I don't understand this change: Haven't you already picked out `$schema` from the spec above? I don't see how `_resolve()` can alter this part of `$spec`. Also, what about fetching `version()` from `SPECIFICATION_URL`?

Reply:
We can extract `'$schema'` from the spec here (line 158) only if the spec was a hashref. If it was provided as a URL, we don't get to see inside it until after we've run _resolve, so when that's returned, then we can look at `$spec->{'$schema'}` and set it if it wasn't already.  _resolve is where the spec is downloaded (or loaded from a local file).